_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,244 @@
+
+The type of metrics exposed by the metric reporters, whether it be Yammer metrics, Kafka metrics, or both, is solely dependent on the implementation of the metric reporter. For example, the JMX metric reporter, a metric reporter provided and automatically enabled by Kafka, only wraps **Kafka metrics** as JMX MBeans. This is counterintuitive since it's known that Kafka exposes all of its metrics, Kafka *and* Yammer, through JMX. The reason why the Kafka JMX metric reporter doesn't wrap **Yammer metrics** as MBeans is because they are already exposed as JMX MBeans by Kafka by default.
+
+Upon startup, every Kafka server loops through a list of metric reporters supplied via its Kafka config, executing every metric reporter on its own thread. As Kafka creates metric objects it passes them to every metric reporter thread for filtering and tracking. What, where, and how this happens to the metric objects depends on the specific implementation of the metric reporter. Afterwards, the metric reporters monitor the metric objects for stat updates from Kafka, formatting and storing the changes for future export.
+
+## Cruise Control Metric Reporter
+Cruise Control ships its own implementation of a Kafka metric reporter, the 'Cruise Control metric reporter'. Unlike the JMX metric reporter, the Cruise Control metric reporter filters the metric objects it receives from Kafka, stowing the metric objects related to Kafka partition monitoring into a concurrent hash map. Then, at user-defined intervals, the reporter will loop through the concurrent hash map and create Cruise Control metric objects from the data provided by the metric objects. For the Cruise Control metric objects, the reporter will include information related to the metric type, value, time of recording, and broker origin. Finally, the reporter will format the Cruise Control metric object into a byte array and store it into a Kafka topic designated for Cruise Control metrics.
+
+Placing the Cruise Control metric reporter jar into the '/opt/kafka/libs/' directory of every Kafka broker allows Kafka to find the reporter at runtime. For the sake of simplicity, one can do this by updating the Strimzi Kafka image with the needed Cruise Control jar using the following Dockerfile:
+
+```Dockerfile
+FROM strimzi/kafka:0.13.0-kafka-2.3.0
+
+ENV CC_VERSION=2.0.57
+ENV JAVA_HOME="/usr/lib/jvm/java-1.8.0-openjdk"
+
+USER root
+
+RUN yum -y install git java-1.8.0-openjdk-devel \
+    && yum clean all -y
+
+RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git
+
+RUN cd cruise-control \
+    && ./gradlew jar :cruise-control-metric-reporter:jar \
+    && cp cruise-control-metrics-reporter/build/libs/* /opt/kafka/libs/
+```
+Then building the image:
+```bash
+docker build . -t <kafka-image-name>
```

Comment:
This will nto work for most people with Minikube etc. We should at least mention that they will need to push it somewhere etc.